From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: notstevy <notstevy@ultrabuildmc.de>
Date: Fri, 26 Sep 2025 16:42:43 +0200
Subject: [PATCH] Predictable Enchantments

Modifying enchantments (such as Unbreaking) to not use random, making pvp more deterministic

Signed-off-by: notstevy <notstevy@ultrabuildmc.de>

diff --git a/net/minecraft/world/item/enchantment/Enchantment.java b/net/minecraft/world/item/enchantment/Enchantment.java
index 0d42a7248e59cc05181da6e9918acf38dab0ac4f..2f8587467acbec315b73135d60625e0bb115d0d4 100644
--- a/net/minecraft/world/item/enchantment/Enchantment.java
+++ b/net/minecraft/world/item/enchantment/Enchantment.java
@@ -212,7 +212,7 @@ public record Enchantment(Component description, Enchantment.EnchantmentDefiniti
 
         for (ConditionalEffect<EnchantmentValueEffect> conditionalEffect : this.getEffects(EnchantmentEffectComponents.DAMAGE_PROTECTION)) {
             if (conditionalEffect.matches(lootContext)) {
-                damageProtection.setValue(conditionalEffect.effect().process(enchantmentLevel, entity.getRandom(), damageProtection.floatValue()));
+                damageProtection.setValue(conditionalEffect.effect().process(enchantmentLevel, entity.getRandom(), damageProtection.floatValue(), level)); // kitpvp - Predicatable enchantments
             }
         }
     }
@@ -374,7 +374,7 @@ public record Enchantment(Component description, Enchantment.EnchantmentDefiniti
         applyEffects(
             this.getEffects(component),
             itemContext(level, enchantmentLevel, tool),
-            enchantmentValueEffect -> value.setValue(enchantmentValueEffect.process(enchantmentLevel, level.getRandom(), value.floatValue()))
+            enchantmentValueEffect -> value.setValue(enchantmentValueEffect.process(enchantmentLevel, level.getRandom(), value.floatValue(), level))  // kitpvp - Predicatable enchantments
         );
     }
 
@@ -389,7 +389,7 @@ public record Enchantment(Component description, Enchantment.EnchantmentDefiniti
         applyEffects(
             this.getEffects(component),
             entityContext(level, enchantmentLevel, entity, entity.position()),
-            enchantmentValueEffect -> value.setValue(enchantmentValueEffect.process(enchantmentLevel, entity.getRandom(), value.floatValue()))
+            enchantmentValueEffect -> value.setValue(enchantmentValueEffect.process(enchantmentLevel, entity.getRandom(), value.floatValue(), level)) // kitpvp - Predicatable enchantments
         );
     }
 
@@ -405,7 +405,7 @@ public record Enchantment(Component description, Enchantment.EnchantmentDefiniti
         applyEffects(
             this.getEffects(component),
             damageContext(level, enchantmentLevel, entity, damageSource),
-            enchantmentValueEffect -> value.setValue(enchantmentValueEffect.process(enchantmentLevel, entity.getRandom(), value.floatValue()))
+            enchantmentValueEffect -> value.setValue(enchantmentValueEffect.process(enchantmentLevel, entity.getRandom(), value.floatValue(), level)) // kitpvp - Predicatable enchantments
         );
     }
 
diff --git a/net/minecraft/world/item/enchantment/EnchantmentHelper.java b/net/minecraft/world/item/enchantment/EnchantmentHelper.java
index b61e0e179e78f37f670ad86476b5a533e7835948..15f2f2743d8fd19b1e2e3c984723a49bf365ccad 100644
--- a/net/minecraft/world/item/enchantment/EnchantmentHelper.java
+++ b/net/minecraft/world/item/enchantment/EnchantmentHelper.java
@@ -346,7 +346,7 @@ public class EnchantmentHelper {
                             if (targetedConditionalEffect.enchanted() == EnchantmentTarget.VICTIM
                                 && targetedConditionalEffect.affected() == EnchantmentTarget.VICTIM
                                 && targetedConditionalEffect.matches(lootContext)) {
-                                mutableFloat.setValue(targetedConditionalEffect.effect().process(level1, random, mutableFloat.floatValue()));
+                                mutableFloat.setValue(targetedConditionalEffect.effect().process(level1, random, mutableFloat.floatValue(), level)); // kitpvp - Predicatable enchantments
                             }
                         }
                     );
@@ -364,7 +364,7 @@ public class EnchantmentHelper {
                                 if (targetedConditionalEffect.enchanted() == EnchantmentTarget.ATTACKER
                                     && targetedConditionalEffect.affected() == EnchantmentTarget.VICTIM
                                     && targetedConditionalEffect.matches(lootContext)) {
-                                    mutableFloat.setValue(targetedConditionalEffect.effect().process(level1, random, mutableFloat.floatValue()));
+                                    mutableFloat.setValue(targetedConditionalEffect.effect().process(level1, random, mutableFloat.floatValue(), level)); // kitpvp - Predicatable enchantments
                                 }
                             }
                         );
diff --git a/net/minecraft/world/item/enchantment/effects/AllOf.java b/net/minecraft/world/item/enchantment/effects/AllOf.java
index c27a372be28b48a54297bdca6839ab177ac7a1f7..c681367c144741e2746bdf7c44360ce80d2c2a73 100644
--- a/net/minecraft/world/item/enchantment/effects/AllOf.java
+++ b/net/minecraft/world/item/enchantment/effects/AllOf.java
@@ -74,6 +74,17 @@ public interface AllOf {
     public record ValueEffects(List<EnchantmentValueEffect> effects) implements EnchantmentValueEffect {
         public static final MapCodec<AllOf.ValueEffects> CODEC = AllOf.codec(EnchantmentValueEffect.CODEC, AllOf.ValueEffects::new, AllOf.ValueEffects::effects);
 
+        // kitpvp start - Predicatable enchantments
+        @Override
+        public float process(int enchantmentLevel, RandomSource random, float value, ServerLevel level) {
+            for (EnchantmentValueEffect enchantmentValueEffect : this.effects) {
+                value = enchantmentValueEffect.process(enchantmentLevel, random, value, level);
+            }
+
+            return value;
+        }
+        // kitpvp start - Predicatable Enchantments
+
         @Override
         public float process(int enchantmentLevel, RandomSource random, float value) {
             for (EnchantmentValueEffect enchantmentValueEffect : this.effects) {
diff --git a/net/minecraft/world/item/enchantment/effects/EnchantmentValueEffect.java b/net/minecraft/world/item/enchantment/effects/EnchantmentValueEffect.java
index 2e5ec28d87c1c2166cb6d899e04fb93d3c91e07c..6d87c23124e1aa491b182206eca4ec9ebbf74c2e 100644
--- a/net/minecraft/world/item/enchantment/effects/EnchantmentValueEffect.java
+++ b/net/minecraft/world/item/enchantment/effects/EnchantmentValueEffect.java
@@ -21,7 +21,14 @@ public interface EnchantmentValueEffect {
         return Registry.register(registry, "set", SetValue.CODEC);
     }
 
-    float process(int enchantmentLevel, RandomSource random, float value);
+    // kitpvp start - Predicatable enchantments
+    default float process(int enchantmentLevel, RandomSource random, float value) {
+        return value;
+    }
+    default float process(int enchantmentLevel, RandomSource random, float value, net.minecraft.server.level.ServerLevel level) {
+        return this.process(enchantmentLevel, random, value);
+    }
+    // kitpvp end - Predicatable enchantments
 
     MapCodec<? extends EnchantmentValueEffect> codec();
 }
diff --git a/net/minecraft/world/item/enchantment/effects/RemoveBinomial.java b/net/minecraft/world/item/enchantment/effects/RemoveBinomial.java
index 8d1fef7736935b826c0c50807c3ea6400f739be8..30fb5881657065776fa209a407cc42b1f1301b48 100644
--- a/net/minecraft/world/item/enchantment/effects/RemoveBinomial.java
+++ b/net/minecraft/world/item/enchantment/effects/RemoveBinomial.java
@@ -11,8 +11,13 @@ public record RemoveBinomial(LevelBasedValue chance) implements EnchantmentValue
     );
 
     @Override
-    public float process(int enchantmentLevel, RandomSource random, float value) {
+    // kitpvp start - Predicatable enchantments
+    public float process(int enchantmentLevel, RandomSource random, float value, net.minecraft.server.level.ServerLevel serverLevel) {
         float f = this.chance.calculate(enchantmentLevel);
+        if(serverLevel.getGameRules().get(world.kitpvp.slime.server.KitPvPGameRulesImpl.PREDICTABLE_ENCHANTMENTS)) {
+            return value - value * f;
+        }
+        // kitpvp end - Predicatable enchantments
         int i = 0;
         if (!(value <= 128.0F) && !(value * f < 20.0F) && !(value * (1.0F - f) < 20.0F)) {
             double floor = Math.floor(value * f);
